macOS: Add support for quick terminal sizing configuration

freepicheep opened the pull request:
Resolves #2384 

Since support for customizing the initial size for the quick terminal has already been added for linux, I just had to expose the configuration to Swift with the C bindings. 

Through the new `QuickTerminalSize` struct the user's custom configuration is used to set the height and width of the quick terminal. It supports pixels or percentage. It also supports a custom width for the quick terminal through the configuration. 

---

EDIT to Include Demos

**Adjusting Height**

![adjusting height](https://github.com/user-attachments/assets/3706fa68-6892-4776-acd8-84f2f635c1a2)

**Adjusting Height and Width**

![adjusting height and width](https://github.com/user-attachments/assets/ce4ad872-3411-4f25-bfc3-ffe41b45caa8)

freepicheep committed:
macOS: Add support for quick terminal sizing configuration

Added C bindings for the already existing quick-terminal-size
configuration. Created a new QuickTerminalSize struct to hold these
values in Swift. Updated the QuickTerminal implementation to use the
user's configuration if supplied. Retains defaults. Also adds support to
customize the width of the quick terminal (height if quick terminal is
set to right or left).

pluiedev reviewed (changes requested):
Only going to comment on the Zig and C parts as I don't know enough about Swift

freepicheep committed:
macOS: update zig and c structs for quick terminal size

Applying the feedback given by @pluiedev to use an enum to specify the
type of quick terminal size configuration given (pixels or percentage).
Updated the Swift code to work with the enum as well.

freepicheep commented:
I broke some things with manually resizing the quick terminal window. I am working on fixing that.

freepicheep committed:
macOS: enable quick terminal manual resizing

You can now resize the quick terminal both vertically and horizontally. To incorporate adjusting the custom secondary size on the quick terminal we needed to have the ability to resize the width (if from top, bottom, or center), and height (if from right, left, or center). The quick terminal will retain the user's manually adjusted size while the app is open. A new feature with this is that when the secondary size is adjusted (or primary if the quick terminal is center), the size will increase or decrease on both sides of the terminal.

freepicheep committed:
Merge branch 'main' into quick-term-initial-size

pluiedev commented on src/config/Config.zig:
```diff
@@ -6234,6 +6234,44 @@ pub const QuickTerminalSize = struct {
+
+        pub const Type = enum(u8) { none, percentage, pixels };
+
+        fn none() CSize {
+            return .{ .type = .none, .value = 0 };
+        }
```
Use decl literals

```suggestion
        pub const none: CSize = .{ .type = .none, .value = 0 };
```

pluiedev commented on src/config/Config.zig:
```diff
@@ -6234,6 +6234,44 @@ pub const QuickTerminalSize = struct {
+
+    pub fn cval(self: QuickTerminalSize) C {
+        return .{
+            .primary = if (self.primary) |p| switch (p) {
+                .percentage => |v| CSize.percentage(v),
+                .pixels => |v| CSize.pixels(v),
+            } else CSize.none(),
+            .secondary = if (self.secondary) |s| switch (s) {
+                .percentage => |v| CSize.percentage(v),
+                .pixels => |v| CSize.pixels(v),
+            } else CSize.none(),
```
Use decl literals

```suggestion
            .primary = if (self.primary) |p| switch (p) {
                .percentage => |v| .percentage(v),
                .pixels => |v| .pixels(v),
            } else .none,
            .secondary = if (self.secondary) |s| switch (s) {
                .percentage => |v| .percentage(v),
                .pixels => |v| .pixels(v),
            } else .none,
```

freepicheep committed:
use decl literals as suggested

Applied from the code review

Co-authored-by: Leah Amelia Chen <github@acc.pluie.me>

freepicheep committed:
macOS: Round quick terminal window position coordinates

This resolves an issue where the right side of the quick terminal would
not resize equally to the left side if adjusting the width from the left
side.

freepicheep reviewed (commented):
I believe the requested changes have been implemented. Thanks for your feedback!

mitchellh commented on include/ghostty.h:
```diff
@@ -445,6 +445,23 @@ typedef struct {
+typedef struct {
+  ghostty_quick_terminal_size_e type;
+  uint32_t value;
+} ghostty_quick_terminal_size_u;
```
This should be suffixed `_s` since its a struct, not a union.

mitchellh commented on macos/Sources/Features/QuickTerminal/QuickTerminalSize.swift:
```diff
@@ -0,0 +1,84 @@
+import GhosttyKit
+
+struct QuickTerminalSize {
```
We should put this in the `Ghostty` folder because its a Swift structure wrapper around a libghostty type.

mitchellh reviewed (changes requested):
I need to review the Swift in more detail but here's some quick feedback

mitchellh commented:
I pushed a rebase, no other changes.

freepicheep committed:
macOS: rename c struct, relocate QuickTerminalSize file

Renamed the ghostty_quick_terminal_size_u to
ghostty_quick_terminal_size_s and moved the QuickTerminalSize file to
the Ghostty folder as requested.

freepicheep commented:
Thanks for your feedback and for rebasing it. I just implemented your suggestions and pushed the update.

freepicheep committed:
Merge branch 'ghostty-org:main' into quick-term-initial-size

freepicheep commented:
Is there anything I can do to move this PR forward?

luweiCN commented:
> I really need this feature. Currently, I’m using Aerospace to open Ghostty with a single hotkey. Before switching to Ghostty, I had been using iTerm2 for a long time, mainly because of its Hotkey Window feature, which allowed me to summon a terminal from any display and any desktop space instantly. I usually ran Neovim inside that window.
> 
> After switching to Ghostty, I was pleasantly surprised to find that it also has a similar feature — Quick Terminal. However, I encountered one limitation that prevents it from fully replacing my previous workflow: the Quick Terminal does not remember its window size.
> 
> Currently, while the Quick Terminal size can be adjusted, it always resets to the default size after I hide and reopen it. Even restarting Ghostty does not preserve the size.
> 
> Suggestions I think having a configurable default size could be nice, but it’s not essential. What’s more important is that the Quick Terminal should remember its last size — even after being hidden, and ideally even after restarting Ghostty.
> 
> I noticed there has been some consideration about which direction the Quick Terminal should appear from, and how the animation and resize behavior work depending on that direction. In my opinion, these are less important. The core need for Quick Terminal users is to quickly summon a terminal from any display and any desktop space using a global hotkey — without interfering with the rest of the workspace. A semi-transparent background, for example, allows you to see the web page or messages underneath.
> 
> Animations and directional controls can be nice to have, but if they make implementation more complex, I’d prefer focusing on the core functionality — remembering window size and being able to summon it from anywhere with a global hotkey.
> 
> I’m not familiar with Zig, so I can’t contribute code to this feature, but I hope my feedback helps clarify the main use case. Really looking forward to seeing this PR merged.

Sorry my mistake, I just found nightly version quick terminal can remember the window size

pluiedev reviewed (approved):
Ack Zig & C changes

mitchellh commented:
I think I remember this having some bugs which is why I didn’t merge after my original feedback was addressed. I’ll have to double check.

mitchellh commented:
Thanks again! I reviewed this, made some changes, and pushed them as #8402 because I couldn't modify this PR. Original authorship was maintained. I'll merge that once CI passes.
